### ISSUE TITLE
SSH transport hangs for big stdout content

### ISSUE THREAD

**Issue by giovannipizzi** [aiidateam/aiida-core#1525]
If a command run through the ssh transport has a very big output (tested with a SGE plugin that, with a few thousands queued jobs, produces an XML output of ~2.5 MB), the SSH hangs forever.
Apparently, this is now mentioned also in the paramiko documentation, see the warning at http://docs.paramiko.org/en/2.4/api/channel.html#paramiko.channel.Channel.recv_exit_status.

A fix that seems to work (tested and solved the issue, giving correct content of the stdout) is the following:
```udiff
diff --git a/aiida/transport/plugins/ssh.py b/aiida/transport/plugins/ssh.py
index 383535408..d02eb8dc3 100644
--- a/aiida/transport/plugins/ssh.py
+++ b/aiida/transport/plugins/ssh.py
@@ -1290,7 +1290,7 @@ class SshTransport(aiida.transport.Transport):
 
         return stdin, stdout, stderr, channel
 
-    def exec_command_wait(self, command, stdin=None, combine_stderr=False, bufsize=-1):
+    def exec_command_wait(self, command, stdin=None, combine_stderr=False, bufsize=-1, read_bufsize=1024):
         """
         Executes the specified command and waits for it to finish.
 
@@ -1299,13 +1299,17 @@ class SshTransport(aiida.transport.Transport):
                    file-like object.
         :param combine_stderr: (optional, default=False) see docstring of
                    self._exec_command_internal()
-        :param bufsize: same meaning of paramiko.
+        :param bufsize: same meaning of paramiko. It is used as the bufsize for the channel files.
+        :param read_bufsize: size in bytes for reading/receiving data from stdout using
+            the channel recv method. This is needed to avoid hangs when there is a very
+            long output (see the warning here: 
+            http://docs.paramiko.org/en/2.4/api/channel.html#paramiko.channel.Channel.recv_exit_status).
+            It must be a positive number, if too big it will hang for big files, if too
+            small it can give performance issues.
 
         :return: a tuple with (return_value, stdout, stderr) where stdout and stderr
             are strings.
         """
-        # TODO: To see if like this it works or hangs because of buffer problems.
-
         ssh_stdin, stdout, stderr, channel = self._exec_command_internal(command, combine_stderr, bufsize=bufsize)
 
         if stdin is not None:
@@ -1325,11 +1329,23 @@ class SshTransport(aiida.transport.Transport):
         ssh_stdin.flush()
         ssh_stdin.channel.shutdown_write()
 
+        # I need to start fetching data in chunks (here I use a )
+        output_text_list = []
+        while True:
+            this_buf = channel.recv(read_bufsize)
+            if not this_buf:
+                # Length zero: it means the channel has closed
+                break
+            output_text_list.append(this_buf)
+
         # I get the return code (blocking)
         retval = channel.recv_exit_status()
 
         # needs to be after 'recv_exit_status', otherwise it might hang
-        output_text = stdout.read()
+        output_text = "".join(output_text_list)
+
+        # TODO: To see if like this it works or hangs because of buffer problems
+        #       and the same thing needs to be done as for stdout       
         stderr_text = stderr.read()
 
         return retval, output_text, stderr_text
```

Probably it needs a bit of testing, but then it can be integrated in the plugin. 
Thought is required to check if the same can happen also with big stderrs (actually I think it would, and it is potentially easy to test probably - just `cat` a big file) but I'm not sure how to use `recv` also for stderr, and how to interleave them correctly.

**Comment by giovannipizzi** [comment]
Probably it's good to add a test for the transport (create a big temp file, then cat it (or cat >&2) and see if it does not hang - here a timeout becomes necessary though to avoid the tests hanging).

**Comment by giovannipizzi** [comment]
Check here for e.g. a much more complex implementation: https://github.com/radssh/radssh/blob/512e532f35e04fccc13fa2df7fd52e06e31df451/radssh/ssh.py

Also, note that there is a recv_stderr function

**Comment by giovannipizzi** [comment]
Check also this https://stackoverflow.com/questions/23504126/do-you-have-to-check-exit-status-ready-if-you-are-going-to-check-recv-ready?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa

**Comment by ConradJohnston** [comment]
@giovannipizzi , Any idea what the current thinking on this issue is?

 > Check also this https://stackoverflow.com/questions/23504126/do-you-have-to-check-exit-status-ready-if-you-are-going-to-check-recv-ready

From your link and also the discussion in [this](https://github.com/paramiko/paramiko/issues/448) issue, is it enough to just change the ordering of operations like this?:

With the current 
```diff
--- a/aiida/transports/plugins/ssh.py
+++ b/aiida/transports/plugins/ssh.py
@@ -1254,13 +1254,13 @@ class SshTransport(Transport):  # pylint: disable=too-many-public-methods
         ssh_stdin.flush()
         ssh_stdin.channel.shutdown_write()

-        # I get the return code (blocking)
-        retval = channel.recv_exit_status()
-
         # needs to be after 'recv_exit_status', otherwise it might hang
         output_text = stdout.read().decode('utf-8')
         stderr_text = stderr.read().decode('utf-8')

+        # I get the return code (blocking)
+        retval = channel.recv_exit_status()
+
         return retval, output_text, stderr_text
```

This prevents hangs (occurring to me in the wild) but I'm not sure if will cause upset later.
The function still waits for an exit code, and as I understand the data retrieved for stdout/err is buffered anyway, and so you can already start capturing it before you know the success of the command.

**Comment by giovannipizzi** [comment]
Hi @ConradJohnston thanks for the message and the links.
Apparently moving down the call might help (I understand from [this issue of Paramiko](https://github.com/paramiko/paramiko/issues/448)) that the behaviour might have changed in the past few years, and this function was written quite some time ago.
However, I am not sure this is enough. See for instance the more complex implementation in the answer of the [stackoverflow question already mentioned above](https://stackoverflow.com/questions/23504126/do-you-have-to-check-exit-status-ready-if-you-are-going-to-check-recv-ready) or the implementation in Fabric, also mentioned in Paramiko issue linked above.

For instance, it's not clear to me what happens if you have continuous large streams *both* on stdout and stderr.

Did you already try to change the code as you suggested? Did it help?
Also, before changing the code I would like to see a test that currently fails and then, after the fix, starts working again - there is a minimal example in the Paramiko issue linked above, that probably can be converted to a AiiDA transport test. Would you be willing to write such a (failing, for now) test and maybe start a PR?

**Comment by ConradJohnston** [comment]
@giovannipizzi , sorry for the latency. 
So I've been running with this change for about 6 weeks now. It did help in the sense that without it, the computer (Archer - UK Tier 1 HPC) wasn't usable. The change allows me to run as normal.  I haven't yet found a case where it breaks, although I haven't been doing any heavy production runs.

I'm happy to have a go at a test and PR, although I can't say when I'll get to it. It's on my Kanban board.

**Comment by giovannipizzi** [comment]
Unfortunately the solution of swapping, while solving the problem for stdout > 2MB (paramiko window size), does not fix the issue fully. First, for larger files, transfer becomes *very* slow.
Secondly, if instead there is a big stderr (rather than stdout), this still hangs (this time while reading the empty stdout: the remote code will hang since no-one is readying stderr, and AiiDA will hang reading all stdout).

I have a different fix, see #3787, that should fix correctly the problem.
Moreover, this also increases speed, for some reason (unclear)!

For instance, using a 0.1s timeout (for the channels), 0.05s sleep (in case of no data received), 10kB buffer, these are the comparisons in time (first number: new implementation in #3787; second number: old implementation, with the only change of putting `retval = channel.recv_exit_status()` below the `read()` to avoid hanging, and `cat`ting a file of given size to stdout):

- 16MB file transfer: 1.2s vs 16s
- 8MB file transfer: 0.7s vs 4.2s
- 4MB file transfer: 0.6s vs 1.4s
- 10kB file transfer: 0.53s vs 0.53s

**Comment by chrisjsewell** [comment]
in the grand tradition of @sphuber lol, I am moving this back a milestone, since (a) I think @giovannipizzi otherwise occupied at present to finish this promptly and, (b) it is not fixing an issue that was not already present in the last release

but obviously let me know if you feel differently
